src/vmm/src/devices/acpi/vmgenid.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use acpi_tables::{aml, Aml};
+use aws_lc_rs::error::Unspecified as RandError;
+use aws_lc_rs::rand;
+use log::{debug, error};
+use serde::{Deserialize, Serialize};
+use utils::eventfd::EventFd;
+use vm_memory::{GuestAddress, GuestMemoryError};
+use vm_superio::Trigger;
+
+use super::super::legacy::EventFdTrigger;
+use crate::device_manager::resources::ResourceAllocator;
+use crate::snapshot::Persist;
+use crate::vstate::memory::{Bytes, GuestMemoryMmap};
+
+/// Virtual Machine Generation ID device
+///
+/// VMGenID is an emulated device which exposes to the guest a 128-bit cryptographically random
+/// integer value which will be different every time the virtual machine executes from a different
+/// configuration file. In Firecracker terms this translates to a different value every time a new
+/// microVM is created, either from scratch or restored from a snapshot.
+///
+/// The device specification can be found here: https://go.microsoft.com/fwlink/?LinkId=260709
+#[derive(Debug)]
+pub struct VmGenId {
+    /// Current generation ID of guest VM
+    pub gen_id: u128,
+    /// Interrupt line for notifying the device about generation ID changes
+    pub interrupt_evt: EventFdTrigger,
+    /// Guest physical address where VMGenID data lives.
+    pub guest_address: GuestAddress,
```

Comment:
instead of guest_address and gsi can we have VMGenIDState here instead?
This will help to pass vmgenid.VMGenIDState from builder.rs for creating vmgenid devicetree node

Reply:
Why can't we just pass  guest_address and gsi, or just a reference to the VMGenID device?

Reply:
I still don't understand why we need a separate object for this, we could just do:

```Rust
        crate::arch::aarch64::configure_system(
            &vmm.guest_memory,
            cmdline,
            vcpu_mpidr,
            vmm.mmio_device_manager.get_device_info(),
            vmm.acpi_manager.vmgenid.as_ref(),
            vmm.vm.get_irqchip(),
            initrd,
        )
```

so, inside the `configure_system` we have an `Option<VmGenID>`, where we can get `gsi` and `guest_addr` from. What problem does a `VMGenIDState` type solve?

Reply:
Thats true, in the example [here](https://github.com/firecracker-microvm/firecracker/commit/f382e09a9b1e4453d2b231232d89bcb85296ac87#diff-a4f542f22c2bab2421c33bb265388e32dddbf0e7bd5d49eb6a4d54c2d17005b4R280) I have done the same. But my point was that along with gsi and guest_addr we are also sending gen_id and interrupt_evt which the file doesn't need. I just feel if the information is not needed by the fdt.rs then it shouldn't receive it.

Reply:
> I just feel if the information is not needed by the fdt.rs then it shouldn't receive it.

What is your concern with that? We are passing a reference to the enclosing type (so we do not have time overhead).

IMHO, it looks an unnecessary complication create a new type for the sole purpose of passing it to the FDT generation code.

Reply:
I think this discussion can be held here: #4542. As of right now, we don't pass acpi info to the fdt. And when we will need to do it, we can think about necessary changes.

Reply:
I agree with Egor. I find it atypical to create a new type just to facilitate "hiding" info for the "potential" case of some coding error. If we really think that's a problem we could just make the fields of VMGenID we are worried about private. But again, I personally do not see any risk about it.

Reply:
discussed offline on the next steps. The VMGenIDState is used for serializing the state during snapshot. Moving it inside vmgenid doesn't provide much benefit. That being said, if we want to protect other members of vmgenid from being passed to fdt we can move them to private but the work is out of scope for this PR.